File: src/main/java/com/flexpoker/table/query/handlers/FlopCardsDealtEventHandler.java
package com.flexpoker.table.query.handlers;

import java.util.Arrays;

import javax.inject.Inject;

import org.springframework.stereotype.Component;

import com.flexpoker.framework.event.EventHandler;
import com.flexpoker.framework.pushnotifier.PushNotificationPublisher;
import com.flexpoker.pushnotifications.TableUpdatedPushNotification;
import com.flexpoker.table.command.events.FlopCardsDealtEvent;
import com.flexpoker.table.query.repository.CardsUsedInHandRepository;
import com.flexpoker.table.query.repository.TableRepository;
import com.flexpoker.web.dto.outgoing.CardDTO;
import com.flexpoker.web.dto.outgoing.TableDTO;

@Component
public class FlopCardsDealtEventHandler implements EventHandler<FlopCardsDealtEvent> {

    private final TableRepository tableRepository;

    private final CardsUsedInHandRepository cardsUsedInHandRepository;

    private final PushNotificationPublisher pushNotificationPublisher;

    @Inject
    public FlopCardsDealtEventHandler(TableRepository tableRepository,
            CardsUsedInHandRepository cardsUsedInHandRepository,
            PushNotificationPublisher pushNotificationPublisher) {
        this.tableRepository = tableRepository;
        this.cardsUsedInHandRepository = cardsUsedInHandRepository;
        this.pushNotificationPublisher = pushNotificationPublisher;
    }

    @Override
    public void handle(FlopCardsDealtEvent event) {
        handleUpdatingTable(event);
        handlePushNotifications(event);
    }

    private void handleUpdatingTable(FlopCardsDealtEvent event) {
        var currentTable = tableRepository.fetchById(event.getAggregateId());

        var flopCards = cardsUsedInHandRepository.fetchFlopCards(event.getHandId());
        var visibleCommonCards = Arrays.asList(
                new CardDTO(flopCards.getCard1().getId()),
                new CardDTO(flopCards.getCard2().getId()),
                new CardDTO(flopCards.getCard3().getId()));

        var updatedTable = new TableDTO(currentTable.getId(),
                event.getVersion(), currentTable.getSeats(),
                currentTable.getTotalPot(), currentTable.getPots(),
                visibleCommonCards, currentTable.getCurrentHandMinRaiseToAmount(),
                currentTable.getCurrentHandId());
        tableRepository.save(updatedTable);
    }

    private void handlePushNotifications(FlopCardsDealtEvent event) {
        var pushNotification = new TableUpdatedPushNotification(event.getGameId(), event.getAggregateId());
        pushNotificationPublisher.publish(pushNotification);
    }

}

File: src/main/kotlin/com/flexpoker/table/query/handlers/FlopCardsDealtEventHandler.kt
package com.flexpoker.table.query.handlers

import com.flexpoker.framework.event.EventHandler
import com.flexpoker.framework.pushnotifier.PushNotificationPublisher
import com.flexpoker.pushnotifications.TableUpdatedPushNotification
import com.flexpoker.table.command.events.FlopCardsDealtEvent
import com.flexpoker.table.query.dto.CardDTO
import com.flexpoker.table.query.repository.CardsUsedInHandRepository
import com.flexpoker.table.query.repository.TableRepository
import org.springframework.stereotype.Component
import javax.inject.Inject

@Component
class FlopCardsDealtEventHandler @Inject constructor(
    private val tableRepository: TableRepository,
    private val cardsUsedInHandRepository: CardsUsedInHandRepository,
    private val pushNotificationPublisher: PushNotificationPublisher
) : EventHandler<FlopCardsDealtEvent> {

    override fun handle(event: FlopCardsDealtEvent) {
        handleUpdatingTable(event)
        handlePushNotifications(event)
    }

    private fun handleUpdatingTable(event: FlopCardsDealtEvent) {
        val tableDTO = tableRepository.fetchById(event.aggregateId)
        val (card1, card2, card3) = cardsUsedInHandRepository.fetchFlopCards(event.handId)
        val visibleCommonCards = listOf(CardDTO(card1.id), CardDTO(card2.id), CardDTO(card3.id))
        val updatedTable = tableDTO.copy(version = event.version, visibleCommonCards = visibleCommonCards)
        tableRepository.save(updatedTable)
    }

    private fun handlePushNotifications(event: FlopCardsDealtEvent) {
        val pushNotification = TableUpdatedPushNotification(event.gameId, event.aggregateId)
        pushNotificationPublisher.publish(pushNotification)
    }

}